src/ServiceControl/Operations/BodyStorage/BodyStorageEnricher.cs

```diff
@@ -11,51 +12,103 @@ public class BodyStorageEnricher : ImportEnricher
 
         public override void Enrich(ImportMessage message)
         {
-            if (message.PhysicalMessage.Body == null || message.PhysicalMessage.Body.Length == 0)
+            var bodySize = GetContentLength(message);
+            message.Metadata.Add("ContentLength", bodySize);
+            if (bodySize == 0)
             {
-                message.Metadata.Add("ContentLength", 0);
                 return;
             }
 
-            string contentType;
+            var contentType = GetContentType(message, "text/xml");
+            message.Metadata.Add("ContentType", contentType);
 
-            if (!message.PhysicalMessage.Headers.TryGetValue(Headers.ContentType, out contentType))
+            var stored = TryStoreBody(message, bodySize, contentType);
+            if (!stored)
             {
-                contentType = "text/xml"; //default to xml for now
+                message.Metadata.Add("BodyNotStored", true);
             }
+        }
 
-            message.Metadata.Add("ContentType", contentType);
+        bool TryStoreBody(ImportMessage message, int bodySize, string contentType)
+        {
+            var bodyId = message.MessageId;
+            var stored = false;
+            var bodyUrl = string.Format("/messages/{0}/body", bodyId);
 
-            var bodySize = message.PhysicalMessage.Body.Length;
+            if (ShouldStoreInBodyStorage(message, bodySize, contentType))
+            {
+                bodyUrl = StoreBodyInBodyStorage(message, bodyId, contentType, bodySize);
+                stored = true;
+            }
+            
+            if (ShouldStoreBodyInMessageMetadata(bodySize, contentType))
+            {
+                message.Metadata.Add("Body", Encoding.UTF8.GetString(message.PhysicalMessage.Body));
+                stored = true;
+            }
 
-            var bodyId = message.MessageId;
+            message.Metadata.Add("BodyUrl", bodyUrl);
+
+            return stored;
+        }
 
-            if (message is ImportFailedMessage || bodySize <= Settings.MaxBodySizeToStore)
+        static bool ShouldStoreInBodyStorage(ImportMessage message, int bodySize, string contentType)
+        {
+            if (message is ImportFailedMessage)
             {
-                StoreBody(message, bodyId, contentType, bodySize);  
+                return true;
             }
-            else
+
+            if (bodySize <= Settings.MaxBodySizeToStore && contentType.Contains("binary"))
```

Comment:
Why? If it's non-binary we're going to store it in the message metadata. What benefit to we gain from putting it in body storage too?